### ISSUE TITLE
Binary pixel stream is detected as text

### ISSUE THREAD

**Issue by pombredanne** [audreyr/binaryornot#10]
This is originally from https://github.com/nexB/scancode-toolkit/issues/50
In this archive https://rubygems.org/downloads/chunky_png-1.2.8.gem the file:
chunky_png-1.2.8.gem-extract/data.tar.gz-extract/spec/resources/pixelstream.rgb
is detected as text even though this is clearly a binary data stream.

libmagic file detects it as octet/stream or data which is correct (for instance with this libmagic ctypes binding: https://github.com/nexB/scancode-toolkit/blob/master/src/typecode/magic2.py )

The issue comes from the fact that bytes above 127 are treated as text, and the test file first 1024 bytes are FF. The original Perl binary detection looked for bytes below 127 only and did some extra stuffs.

**Comment by pombredanne** [comment]
@audreyr There are a few possible ways to fix this:

1. change the heuristics to treat bytes below and above 127 differently. Use some new heuristic based on the proportion of high bytes to declare a file as binary. 

2. keep the heuristics as-is and introduce chardet as a dependency. If a file is detected as binary in certain cases (excluding the null byte detection), apply a second level detection using `chardet` to see if an encoding can be guessed with a high enough confidence (say above 0.8). If so try to convert the buffer to unicode using the detected encoding. If this fails the file is binary, else the file is text.

3. mix approach 1 and 2, using 2. only when high bytes are found.

**Comment by pombredanne** [comment]
Note also this for testing for UTF8 https://github.com/etrepum/pyutf8

**Comment by pombredanne** [comment]
And fwiw, if we add chardet to things still detected as binary this way, we would get this:
```
./binaryornot/tests/isBinaryFile/encodings/bom_utf-16.txt: UTF-16BE with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/bom_utf-32le.txt: UTF-32LE with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/test-gb2.txt: GB2312 with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/test-kr.txt: EUC-KR with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/test-latin.txt: MacCyrillic with confidence 0.837785152976
./binaryornot/tests/isBinaryFile/encodings/big5.txt: Big5 with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/test-utf16be.txt: UTF-16BE with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/test-gb.txt: GB2312 with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/bom_utf-32.txt: UTF-32BE with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/bom_utf-8.txt: UTF-8 with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/big5_B.txt: Big5 with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/test-shishi.txt: utf-8 with confidence 0.99
./binaryornot/tests/isBinaryFile/encodings/utf_8.txt: utf-8 with confidence 0.7525
./binaryornot/tests/isBinaryFile/encodings/bom_utf-16le.txt: UTF-16LE with confidence 1.0
./binaryornot/tests/isBinaryFile/encodings/utf8cn.txt: utf-8 with confidence 0.99
```
Then we could try a `unicode` conversion on these to confirm the chardet encoding and decide text or binary for lesser travelled texts encodings.

**Comment by pombredanne** [comment]
This second commit builds on top of the previous to support add `chardet` encoding detection and now reports as text all UTF8/16/32 variants. Only pdf fails to be detected as binary (this is not a regression, your code failed at it before too)

**Comment by pombredanne** [comment]
@audreyr Please check these two commits and tell me what you like best... I will update and create a PR then

**Comment by pydanny** [comment]
Hey @pombredanne, I was looking at your work and while I prefer the heuristics to keep dependencies down, I also see the virtue of using `chardet`. @audreyr will make the final call, of course. :smile: 

Where I am concerned is Windows compatibility. `Binaryornot` is a dependency of `Cookiecutter`, which currently works on all current versions of Python (including PyPy) and works identically across Mac OS X, Linux, and Windows. We don't want to lose that capability, especially for Windows users (who tend to be newer to coding).

Therefore, what I'm going to do is add Windows continuous integration to this repo. I should have that done today, whereupon we can test solutions against a wider range of systems.

**Comment by audreyr** [comment]
@pombredanne I like your mixed approach, amazing work here. I have no problem adding chardet as a dependency.

@pydanny Thanks for offering to work on adding Windows continuous integration. I don't think it would be a problem to keep Windows compatibility.

**Comment by pombredanne** [comment]
@pydanny FWIW, scancode is supported and tested on Linux, Windows and Mac...
I run tests on Travis for Linux 64: https://travis-ci.org/nexB/scancode-toolkit
For windows on appveyor:  https://ci.appveyor.com/project/nexB/scancode-toolkit

So whatever suggestions @audreyr accepts, it will have to be working there for me. Both binaryornot and chardet are already in use there and heavily tested on these 3 oses.

**Comment by pombredanne** [comment]
@audreyr wrote:
> I like your mixed approach, amazing work here. I have no problem adding chardet as a dependency.
Let me transform this in a bona fide pull request.

**Comment by pombredanne** [comment]
@audreyr I assume that you have no issue with the inclusions of the excellent tests files of https://github.com/gjtorikian/isBinaryFile from @gjtorikian ? I included the license notice.

**Comment by audreyr** [comment]
No issue, those are very nice tests. The more test files, the merrier. And isBinaryFile looks great, something I'll be studying for ideas :smile:

**Comment by pombredanne** [comment]
and @audreyr Thanks for the kudos. This means a lot as I hold your's  @pydanny 's code such as cookiecutter and django-packages/opencomparison in high respect!

**Comment by pombredanne** [comment]
@audreyr Do you really want to support Python 2.6?

**Comment by audreyr** [comment]
I could easily be convinced to drop 2.6 if it's too much of a hassle.

**Comment by audreyr** [comment]
Fixed by @pombredanne in #11.
